libs/geometry/src/TangentSpaceMesh.cpp

```diff
@@ -187,14 +171,157 @@ Algorithm selectAlgorithm(TangentSpaceMeshInput *input) {
     return outAlgo;
 }
 
-#undef IS_INPUT_TYPE
+// The paper uses a Z-up world basis, which has been converted to Y-up here
+void frisvadKernel(const float3& N, float3& T, float3& B) noexcept {
+    if (N.y < -1.0f + std::numeric_limits<float>::epsilon()) {
+        // Handle the singularity
+        T = float3{-1.0f, 0.0f, 0.0f};
+        B = float3{0.0f, 0.0f, -1.0f};
+        return;
+    }
+    const float a = 1.0f / (1.0f + N.y);
+    const float b = -N.z * N.x * a;
+    T = float3{b, -N.z, 1.0f - N.z * N.z * a};
+    B = float3{1.0f - N.x * N.x * a, -N.x, b};
+}
+
+void hughesMollerKernel(const float3& N, float3& T, float3& B) noexcept {
+    if (abs(N.x) > abs(N.z) + std::numeric_limits<float>::epsilon()) {
+        T = float3{-N.y, N.x, 0.0f};
+    } else {
+        T = float3{0.0f, -N.z, N.y};
+    }
+    T = normalize(T);
+    B = cross(N, T);
+}
+
+void normalsOnlyMethod(const TangentSpaceMeshInput* input, TangentSpaceMeshOutput* output,
```

Comment:
I'm wondering if we shouldn't pass `NormalsOnlyKernelPtr` as a template parameter instead so that we don't have a function call in the main loop. Ideally we want the main loop to be all inlined and possible vectorized.

Reply:
Done.  Folded all the function calls into the loop.